contracts/dev/ocr2dr/OCR2DRClient.sol

```diff
@@ -0,0 +1,129 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.6;
+
+import "./OCR2DR.sol";
+import "./interfaces/OCR2DRClientInterface.sol";
+import "./interfaces/OCR2DROracleInterface.sol";
+
+/**
+ * @title The OCR2DR client contract
+ * @notice Contract writers can inherit this contract in order to create on-demand OCR requests
+ */
+abstract contract OCR2DRClient is OCR2DRClientInterface {
+    OCR2DROracleInterface private s_oracle;
+    mapping(bytes32 => address) private s_pendingRequests;
+
+    event RequestSent(bytes32 indexed id);
+    event RequestFulfilled(bytes32 indexed id);
+
+    error SenderIsNotOracle();
+    error RequestIsAlreadyPending();
+    error RequestIsNotPending();
+
+    constructor(address oracle) {
+        setOracle(oracle);
+    }
+
+    function getDONPublicKey() external view override returns (bytes32) {
+        return s_oracle.getDONPublicKey();
+    }
+
+    /**
+     * @notice Sends OCR2DR request to the stored oracle address
+     * @param req The initialized OCR2DR.Request
+     * @param subscriptionId The subscription ID
+     * @return requestId The generated request ID
+     */
+    function sendRequest(OCR2DR.Request memory req, uint256 subscriptionId)
+        internal
+        returns (bytes32)
+    {
+        bytes32 requestId = s_oracle.sendRequest(subscriptionId, OCR2DR.encodeCBOR(req));
+        s_pendingRequests[requestId] = address(s_oracle);
+        emit RequestSent(requestId);
+        return requestId;
+    }
+
+    /**
+     * @notice User defined function to handle a response
+     * @param requestId The request ID, returned by sendRequest()
+     * @param response Aggregated response from the user code
+     * @param err Aggregated error from the user code or from the execution pipeline
+     * Either response or error parameter will be set, but never both
+     */
+    function fulfillRequest(
+        bytes32 requestId,
+        bytes memory response,
+        bytes memory err
+    ) internal virtual;
+
+    /**
+     * @notice OCR2DR response handler called by the designated oracle
+     * @dev This is a proxy function to user-defined fulfillRequest(),
+     * used to validate and record a fulfillment before calling the user's callback
+     * @param requestId Same as for fulfillRequest()
+     * @param response Same as for fulfillRequest()
+     * @param err Same as for fulfillRequest()
+     */
+    function handleOracleFulfillment(
+        bytes32 requestId,
+        bytes memory response,
+        bytes memory err
+    ) external override recordChainlinkFulfillment(requestId) {
+        fulfillRequest(requestId, response, err);
+    }
```

Comment:
Fine for v1. Probably something to think about for v2, since this proxy does cost extra gas.